### ISSUE TITLE
Error with BS4 and orphan = true

### ISSUE THREAD

**Issue by Sell24** [IGreatlyDislikeJavascript/bootstrap-tourist#37]
Versions
BS4: 4.0.0
JQ: 3.4.1
Popper: 1.12.9

When I try to set a step as an orphan:

                                {
                                    orphan: true,
                                    title: "Title",
                                    content: "Content",
                                },   

I get the following error: 

util.js:146 Uncaught Error: TOOLTIP: Option "offset" provided type "function" but expected type "(number|string)".

Since I am using this for the first time, this is the correct way to generate a popup without tying it to an element correct?

**Comment by IGreatlyDislikeJavascript** [comment]
Hi Sell24, thanks for your feedback - I'll comment on your other issues when I get in front of the code again later.

For this issue, have you specified the `framework: "bootstrap4"` option in the global tour options? If so, I wonder if something has changed between the BS4.0.0 you're using, and current versions that's causing this. The next step I'd suggest is trying it with a current BS and Popper version, as you're several revisions behind the current.

Yes, your snippet is the correct way to create that unattached tour step - we're calling that an "intended orphan".

**Comment by Sell24** [comment]
Sorry thought I did update to the latest, but you were right I wasnt up to date. I copied the script includes directly from the Bootstrap 4 website so I believe I am now up to date. But I am still getting the same error. So here is my latest info:

BS4: 4.1.13
JQ: 3.4.1
Popper: 1.14.3
Browser: Chrome Version 77.0.3865.90 (Official Build) (64-bit)
Framework: VueJS

To answer your question, yes I am using the framework: "bootstrap4" option. The popup works great if it is not an orphan but generates the error if it is an orphan. And just in case you missed it in my other issues, I am using this within the VueJS framework so maybe there is a conflict with it?

I looked at the Popper documentation (no offense to them, but boy is it a hard to follow) and some examples online and in each case they specify offset as a string or number not an object like in your code.

However, you are way more familiar with it and how it is used in your code than I am, so I just wanted to include this link to a thread to Github in regards to popper.js and offsets. Maybe it can help?

[https://github.com/FezVrasta/popper.js/issues/107](url)

And just to be complete here is a full trace of the error from Chrome Dev Tools after updating.

>Uncaught Error: TOOLTIP: Option "offset" provided type "function" but expected type "(number|string)".
    at Object.typeCheckConfig (util.js:137)
    at i.t._getConfig (tooltip.js:632)
    at i (tooltip.js:129)
    at new i (popover.js:11)
    at HTMLBodyElement.<anonymous> (popover.js:158)
    at Function.each (jquery-3.4.1.min.js:2)
    at k.fn.init.each (jquery-3.4.1.min.js:2)
    at k.fn.init.i._jQueryInterface [as popover] (popover.js:149)
    at Tour._showPopover (bootstrap-tourist.js?d395:1859)
    at Tour._showPopoverAndOverlay (bootstrap-tourist.js?d395:1711)

Again thanks for help and the quick response.

**Comment by IGreatlyDislikeJavascript** [comment]
Ok I'll need to look into this, I'm not sure why you're getting that error. Maybe it's to do with your framework, I test Tourist against latest bootstrap 4 with popper.js bundled, perhaps there's a difference somewhere.

Whilst I try and figure out the source or next steps, would you be able to quickly try the unstable/new-backdrop branch please?

You'll need both the css and js from that branch, it is unstable for the moment but will help me work out if the problem is solved via the changes in that fork.

Thanks!

**Comment by Sell24** [comment]
Sure no issues. Will test it out and report back. 

Just as an aside, is it possible that the version of tourist I downloaded is not complete given the missing variable declarations as in my two previous issues? It is labelled v0.2.0 but maybe something happened?

**Comment by Sell24** [comment]
I tried it and same error. And just an FYI, those rslt variables are still undeclared.

I also realized I made a mistake on where the error is occurring. The code is referencing popover.js and not popper.js directly. I apologize for the mistake.

I will continue to look at my code and see if there is something causing the conflict.

Thanks again.

**Comment by Sell24** [comment]
Some more info:
I debugged into the code and in tooltip.js there is this declaration: 

>   const DefaultType = {
    animation           : 'boolean',
    template            : 'string',
    title               : '(string|element|function)',
    trigger             : 'string',
    delay               : '(number|object)',
    html                : 'boolean',
    selector            : '(string|boolean)',
    placement           : '(string|function)',
    offset              : '(number|string)',
    container           : '(string|element|boolean)',
    fallbackPlacement   : '(string|array)',
    boundary            : '(string|element)'
  }

The code does a type check and fails for the offset.  I checked my link address and it is referencing the following: https://stackpath.bootstrapcdn.com/bootstrap/js/src/tooltip.js so it is not pointing to an outdated version.  I even downloaded the BS4 source for 4.1.3 and it has the same declaration.

Any chance you could look at your version of tooltip.js and see if there is a difference with yours?

**Comment by IGreatlyDislikeJavascript** [comment]
> Any chance you could look at your version of tooltip.js and see if there is a difference with yours?

I'm using the bootstrap bundle, although I can't see any reason why there would be a difference I suspect this is the problem.

So whilst I understand what you've posted as the decl for tooltip, it's not the same for Bootstrap popper, which wraps the initialization of popper and only exposes some popper options. The docs for popper are terrible, but you can see the function docs here: https://popper.js.org/popper-documentation.html#modifiers..offset.fn

That link basically says that the popper offset.offset can be a modifier function, and Bootstrap exposes that through the offset option, which is how Tourist shows an orphan tour step. I did have to do all sorts of strange stuff to get popper.js to show an unattached popover in the middle of the screen - it's not as simple as BS3 unfortunately.

You should be able to see the following in your bootstrap:

```
  var NAME$6 = 'tooltip';
  var VERSION$6 = '4.3.1';
  var DATA_KEY$6 = 'bs.tooltip';
  var EVENT_KEY$6 = "." + DATA_KEY$6;
  var JQUERY_NO_CONFLICT$6 = $.fn[NAME$6];
  var CLASS_PREFIX = 'bs-tooltip';
  var BSCLS_PREFIX_REGEX = new RegExp("(^|\\s)" + CLASS_PREFIX + "\\S+", 'g');
  var DISALLOWED_ATTRIBUTES = ['sanitize', 'whiteList', 'sanitizeFn'];
  var DefaultType$4 = {
    animation: 'boolean',
    template: 'string',
    title: '(string|element|function)',
    trigger: 'string',
    delay: '(number|object)',
    html: 'boolean',
    selector: '(string|boolean)',
    placement: '(string|function)',
    offset: '(number|string|function)', <<<<<< accepts a function
    container: '(string|element|boolean)',
    fallbackPlacement: '(string|array)',
    boundary: '(string|element)',
    sanitize: 'boolean',
    sanitizeFn: '(null|function)',
    whiteList: 'object'
  };
```

The key here is that Bootstrap wrapper of popper.js behaves differently to direct use of popper.js. 

My suggestion would be to remove the explicit/separate include of popper.js, and use the latest bootstrap bundle to see if that solves this issue. 

Unfortunately at this moment I'm very sure it's not a problem with the Tourist code.

**Comment by Sell24** [comment]
If you say it works with the bundle then I would agree with you that it is something within my code. I am using an open-source third party admin interface and I am currently going through it and removing all unnecessary add-ins and code that I can find. 

I will look into using the bundle and keep messing around and see if I can find the issue.

Thanks again for your help.

**Comment by Sell24** [comment]
Well I can confirm that switching to the bundle works whereas loading them individually as per the BS4 documentation does not.

I guess I will chalk it up to "whatever". Thanks again for taking the time to answer.

**Comment by ibastevan** [comment]
There is definitely something going on, I can't get orphans working with Bootstrap bundle v4.1.3. 
I updated mine to v4.3.1 and it is now working.
So perhaps we need to state this somewhere as a minimum compatibility?

**Comment by IGreatlyDislikeJavascript** [comment]
I'll add that to the docs. It's got to be due to changes to BS and popper in the intervening versions.
